sdk/data/azcosmos/cosmos_global_endpoint_manager_test.go

```diff
@@ -0,0 +1,407 @@
+
+	expectedLocation := "West US"
+	assert.Equal(t, expectedLocation, location)
+}
+
+func TestGlobalEndpointManagerGetAccountProperties(t *testing.T) {
+	headerPolicy := &headerPolicies{}
+	srv, close := mock.NewTLSServer()
+	defer close()
+	srv.SetResponse(mock.WithStatusCode(http.StatusOK))
+
+	verifier := headerPoliciesVerify{}
+	pl := azruntime.NewPipeline("azcosmostest", "v1.0.0", azruntime.PipelineOptions{PerCall: []policy.Policy{headerPolicy, &verifier}}, &policy.ClientOptions{Transport: srv})
+	req, err := azruntime.NewRequest(context.Background(), http.MethodGet, srv.URL())
+	if err != nil {
+		return
+	}
+	req.SetOperationValue(pipelineRequestOptions{
+		isWriteOperation: true,
+	})
+
+	client := &Client{endpoint: srv.URL(), pipeline: pl}
+
+	preferredRegions := []string{"West US", "Central US"}
+
+	gem, err := newGlobalEndpointManager(client, preferredRegions, 5*time.Minute)
+	assert.NoError(t, err)
+
+	accountProps, err := gem.GetAccountProperties()
+	assert.NoError(t, err)
+
+	expectedAccountProps := accountProperties{
+		ReadRegions:                  nil,
+		WriteRegions:                 nil,
+		EnableMultipleWriteLocations: false,
+	}
+	assert.Equal(t, expectedAccountProps, accountProps)
+}
+
+func TestGlobalEndpointManagerCanUseMultipleWriteLocations(t *testing.T) {
+	headerPolicy := &headerPolicies{}
+	srv, close := mock.NewTLSServer()
+	defer close()
+	srv.SetResponse(mock.WithStatusCode(http.StatusOK))
+
+	verifier := headerPoliciesVerify{}
+	pl := azruntime.NewPipeline("azcosmostest", "v1.0.0", azruntime.PipelineOptions{PerCall: []policy.Policy{headerPolicy, &verifier}}, &policy.ClientOptions{Transport: srv})
+	req, err := azruntime.NewRequest(context.Background(), http.MethodGet, srv.URL())
+	assert.NoError(t, err)
+
+	req.SetOperationValue(pipelineRequestOptions{
+		isWriteOperation: true,
+	})
+
+	client := &Client{endpoint: srv.URL(), pipeline: pl}
+
+	preferredRegions := []string{"West US", "Central US"}
+
+	serverEndpoint, err := url.Parse(srv.URL())
+	assert.NoError(t, err)
+
+	mockLc := newLocationCache(preferredRegions, *serverEndpoint)
+	mockLc.enableMultipleWriteLocations = true
+	mockLc.useMultipleWriteLocations = true
+
+	mockGem := globalEndpointManager{
+		client:              client,
+		preferredLocations:  preferredRegions,
+		locationCache:       mockLc,
+		refreshTimeInterval: 5 * time.Minute,
+	}
+
+	gem, err := newGlobalEndpointManager(client, preferredRegions, 5*time.Minute)
+	assert.NoError(t, err)
+
+	// Multiple locations should be false for default GEM
+	canUseMultipleWriteLocs := gem.CanUseMultipleWriteLocations()
+	assert.False(t, canUseMultipleWriteLocs)
+
+	// Mock GEM with multiple write locations available should show true
+	canUseMultipleWriteLocs = mockGem.CanUseMultipleWriteLocations()
+	assert.True(t, canUseMultipleWriteLocs)
+}
+
+func TestGlobalEndpointManagerUpdate(t *testing.T) { //This test should be testing for the lock on the update/ refresh mechanism
```

Comment:
How is this test checking the lock? Are we issuing multiple concurrent Update calls and measuring how many requests the pipeline did?

Reply:
I left the note to change the test and forgot to do it 😢